src/operator/src/bulk_insert.rs

```diff
@@ -0,0 +1,123 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use ahash::{HashMap, HashMapExt};
+use api::v1::region::{
+    bulk_insert_request, region_request, ArrowIpc, BulkInsertRequest, RegionRequest,
+    RegionRequestHeader, RegionSelection,
+};
+use common_base::AffectedRows;
+use common_grpc::flight::{FlightDecoder, FlightEncoder, FlightMessage};
+use common_grpc::FlightData;
+use common_telemetry::tracing_context::TracingContext;
+use prost::Message;
+use snafu::ResultExt;
+use store_api::storage::RegionId;
+use table::metadata::TableId;
+
+use crate::error;
+use crate::insert::Inserter;
+
+impl Inserter {
+    /// Handle bulk insert request.
+    pub async fn handle_bulk_insert(
+        &self,
+        table_id: TableId,
+        decoder: &mut FlightDecoder,
+        data: FlightData,
+    ) -> error::Result<AffectedRows> {
+        let raw_flight_data = data.encode_to_vec();
+        // Build region server requests
+        let message = decoder
+            .try_decode(data)
+            .context(error::DecodeFlightDataSnafu)?;
+        let FlightMessage::Recordbatch(rb) = message else {
+            return Ok(0);
+        };
+
+        // todo(hl): find a way to embed raw FlightData messages in greptimedb proto files so we don't have to encode here.
+        // safety: when reach here schema must be present.
+        let schema_message = FlightEncoder::default()
+            .encode(FlightMessage::Schema(decoder.schema().unwrap().clone()));
+        let schema_data = schema_message.encode_to_vec();
+
+        let record_batch = rb.df_record_batch();
+        let partition_rule = self
+            .partition_manager
+            .find_table_partition_rule(table_id)
+            .await
+            .context(error::InvalidPartitionSnafu)?;
+
+        let mut handles = Vec::new();
+        // find partitions for each row in the record batch
+        let region_masks = partition_rule.split_record_batch(record_batch).unwrap();
+
+        let mut mask_per_datanode = HashMap::with_capacity(region_masks.len());
+        for (region_number, mask) in region_masks {
+            let region_id = RegionId::new(table_id, region_number);
+            let datanode = self
+                .partition_manager
+                .find_region_leader(region_id)
+                .await
+                .context(error::FindRegionLeaderSnafu)?;
+            let selection = RegionSelection {
+                region_id: region_id.as_u64(),
+                selection: mask.values().inner().as_slice().to_vec(),
+            };
+            mask_per_datanode
+                .entry(datanode)
+                .or_insert_with(Vec::new)
+                .push(selection);
+        }
+
+        for (peer, masks) in mask_per_datanode {
+            let node_manager = self.node_manager.clone();
+            let schema = schema_data.clone();
+            let payload = raw_flight_data.clone();
```

Comment:
I think it's worth a fast path for single region situation, to avoid the clone here.